### ISSUE TITLE
add UNIX domain socket support

### ISSUE THREAD

**Issue by david415** [tikv/grpc-rs#296]
Support for using unix domain socket would be nice.

**Comment by BusyJay** [comment]
Good idea! Would you like to give it a try?

**Comment by david415** [comment]
https://github.com/david415/grpc-rs/tree/unix_socket_listen.0

**Comment by BusyJay** [comment]
Cool, feel free to send us pull request when it's ready.

**Comment by david415** [comment]
i don't have time to make it work.

**Comment by gedigi** [comment]
I have a [diff](https://github.com/gedigi/grpc-rs/commit/bb0eff9bcc186cbfc059d83fc37f8c073309668a) that I don't know why it doesn't work. My test fails in `grpc_server_setup_transport()` line 1178:

```c
  size_t cq_idx;
  for (cq_idx = 0; cq_idx < s->cq_count; cq_idx++) {
    if (grpc_cq_pollset(s->cqs[cq_idx]) == accepting_pollset) break;
  }
  if (cq_idx == s->cq_count) {
    /* completion queue not found: pick a random one to publish new calls to */
    cq_idx = static_cast<size_t>(rand()) % s->cq_count; <--
  }
  chand->cq_idx = cq_idx;
```

`s->cq_count` is 0. I'm not super familiar with native grpc. Any idea why it fails?

**Comment by kate-goldenring** [comment]
Did this ever get resolved? Is there a way to use grpcio for UNIX domain sockets in rust?

**Comment by BusyJay** [comment]
@kate-goldenring @david415 I have implemented it as #446, do you want to give it a try to see if it suits your need?

**Comment by kate-goldenring** [comment]
@BusyJay thanks for working on this. I ended up using tonic grpc instead.

**Comment by AntonioDan** [comment]
Hi, can we use UDS with grpc-rs and is there some example?

**Comment by AntonioDan** [comment]
> Does the test case meet your need?
> 
> https://github.com/tikv/grpc-rs/blob/f897d1d923069b44a54f13d4eded1bfe3c610079/tests-and-examples/tests/cases/misc.rs#L166-L199
> 
> The key point is how to provide socket path.

Yes, I have tried this and it can work. Besides, does grpc-rs support abstract uds?
I tried but meet " BindFail("unix:@/tmp/test_socket", 0)'", not sure if I do anything wrong.

**Comment by BusyJay** [comment]
Unfortunately, grpc-rs can't help with it because grpc c core doesn't support it. See https://github.com/grpc/grpc/issues/4677.

**Comment by BusyJay** [comment]
Abstract UDS support will be fixed in #523.
